File: Algorithms/Sorting/Insertion Sort/insertion_sort.cpp
#include <bits/stdc++.h>
using namespace std;

// Insertion Sort Prototype.
void insertionSort(int[], int);
void printArray(int[], int);

/**
 * main - Program's Entry Point.
 * Description: Implementing C++ Insertion Sort.
 * Return: (0) - Always Success.
 */
int main() 
{
  int arr[] = {2, 4, 5, 3, 1, 6, 7, 9, 8, 1, 0};
  int N = sizeof(arr) / sizeof(arr[0]);
  insertionSort(arr, N);
  cout << "Sorted array: \n";
  printArray(arr, N);
  
  return 0;
}

/**
 * Insertion Sort - Insertion Sort Function.
 * Description: Sorting an Array of Elements (int).
 * @arr: The Array to Sort.
 * @length: The Size of the Array.
 * Return: (void) Nothing.
 */
void insertionSort(int arr[], int length)
{
  for (int i = 0; i < length; i++)
  {
    int currentElement = arr[i]; // Pointer (Indicator) for The Current Element to Compare.
    int j = i - 1; // Indicator for The Next Iteration Process (Could Be a For Loop).
    while (j >= 0 && arr[j] > currentElement) // While Still in Iteration Range & Previous Element is Bigger than Current.
    {
      arr[j + 1] = arr[j]; // Assign The Current Place for The Bigger Value.
      j--;
    }
    /* If The Previous Loop Condition Didn't Met. */
    arr[j + 1] = currentElement; // Then Add The Curent Value to Opposite Direction.
  }
}

// Function to print an array.
void printArray(int arr[], int size)
{
  int i;
  for (i = 0; i < size; i++)
    cout << arr[i] << " ";
  cout << endl;
}

File: Algorithms/Sorting/Insertion Sort/insertionSort.js
let arr = [1, 44, 5, 78, 22, 4, 9, 90];

function sort(arr) {
  /* 
    	Assuming the left portion is a sorted and we'll start from the second element
        as the begging of the second portion (un-sorted array).
    */
  for (let i = 1; i < arr.length; i++) {
    // Indicator for the Current Element.
    let currentElement = arr[i];
    // Indicator for the previous Index.
    let j = i - 1;
    // Comparing and Swapping if Needed.
    while (j >= 0 && currentElement < arr[j]) {
      arr[j + 1] = arr[j];
      j -= 1;
    }
    // Putting the highest element into the last index.
    arr[j + 1] = currentElement;
  }
}

sort(arr);
console.log(arr);
